File: src/lib/app/prefs.ts
/**
 * Local app preferences — the no-login personalization layer.
 *
 * The website personalizes via a subscriber's confirmation_token (emailed,
 * never held by the app). The app therefore keeps its own local preferences in
 * Capacitor Preferences so personalization works for everyone immediately;
 * when the user later subscribes and confirms, the newsletter side personalizes
 * independently. Topic prefs here drive in-app filtering/ordering.
 */
import { cacheGet, cacheSet } from './native';

export type AppPrefs = {
	categories: string[]; // empty = all topics
	hasKids: boolean | null;
	pushWanted: boolean; // user's intent; real APNs registration lands in M2
	email: string | null; // last email the user subscribed with (for display)
	onboarded: boolean;
};

const KEY = 'app_prefs_v1';

const DEFAULTS: AppPrefs = {
	categories: [],
	hasKids: null,
	pushWanted: false,
	email: null,
	onboarded: false
};

export async function loadPrefs(): Promise<AppPrefs> {
	const stored = await cacheGet<Partial<AppPrefs>>(KEY);
	return { ...DEFAULTS, ...(stored ?? {}) };
}

export async function savePrefs(patch: Partial<AppPrefs>): Promise<AppPrefs> {
	const next = { ...(await loadPrefs()), ...patch };
	await cacheSet(KEY, next);
	return next;
}

File: ios-native/NurEine/NurEine/NurEine/Services/Prefs.swift
import Foundation
import SwiftUI

/// Local, no-login preferences (UserDefaults) — topics, onboarding, push intent.
/// Mirrors src/lib/app/prefs.ts.
@Observable
final class Prefs {
    var categories: Set<String> {
        didSet { defaults.set(Array(categories), forKey: "categories") }
    }
    var hasKids: Bool? {
        didSet { defaults.set(hasKids, forKey: "hasKids") }
    }
    var pushWanted: Bool {
        didSet { defaults.set(pushWanted, forKey: "pushWanted") }
    }
    var onboarded: Bool {
        didSet { defaults.set(onboarded, forKey: "onboarded") }
    }
    var email: String? {
        didSet { defaults.set(email, forKey: "email") }
    }

    private let defaults = UserDefaults.standard

    init() {
        categories = Set(defaults.stringArray(forKey: "categories") ?? [])
        hasKids = defaults.object(forKey: "hasKids") as? Bool
        pushWanted = defaults.bool(forKey: "pushWanted")
        onboarded = defaults.bool(forKey: "onboarded")
        email = defaults.string(forKey: "email")
    }
}
